st2common/tests/unit/test_rbac_resolvers_rule_enforcement.py

```diff
@@ -0,0 +1,427 @@
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role with no permission grants, should return false for everything
+        user_db = self.users['1_custom_role_no_permissions']
+        self.assertFalse(self._user_has_resource_db_permissions(
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role with unrelated permission grant to parent pack
+        user_db = self.users['custom_role_pack_grant']
+        self.assertFalse(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_1'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+        self.assertFalse(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_2'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+
+        # Custom role with with grant on the parent pack
+        user_db = self.users['custom_role_rule_pack_grant']
+        self.assertTrue(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_1'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+        self.assertTrue(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_2'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+
+        # Custom role with a direct grant on rule
+        user_db = self.users['custom_role_rule_grant']
+        self.assertTrue(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_3'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+
+        # Custom role - "rule_all" grant on the rule parent pack
+        user_db = self.users['custom_role_pack_rule_all_grant']
+        resource_db = self.resources['rule_enforcement_1']
+        self.assertTrue(self._user_has_resource_db_permissions(
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role - "rule_all" grant on the rule
+        user_db = self.users['custom_role_rule_all_grant']
+        resource_db = self.resources['rule_enforcement_1']
+        self.assertTrue(self._user_has_resource_db_permissions(
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role - "rule_modify" grant on rule_1
+        # XXX: This test is failing because I think there is a bug in
```

Comment:
@Kami @manas: So I think there is an issue that when an user has modify permissions alone on a rule, enforcement view won’t work because https://github.com/StackStorm/st2/blob/master/st2common/st2common/services/rbac.py#L169

https://github.com/StackStorm/st2/pull/2253/files#diff-972afd8e65629a580a1b3b13f3645d2dR538

So when we create a modify permission for a user, do we write a view grant as well? When I dumped the permission grant models from database for this case, I see a modify but not a view grant. I could change the permission_types in the array to include CREATE and MODIFY if the requested type is VIEW but that doesn't seem right. Maybe we do write the VIEW grant in the code but fixtures don't capture it?